core/trino-main/src/main/java/io/trino/execution/scheduler/FaultTolerantStageScheduler.java

```diff
@@ -0,0 +1,529 @@
+            }
+            else {
+                // stage will be consumed by the coordinator using direct exchange
+                outputBuffers = createInitialEmptyOutputBuffers(PARTITIONED)
+                        .withBuffer(new OutputBuffers.OutputBufferId(0), 0)
+                        .withNoMoreBufferIds();
+                exchangeSinkInstanceHandle = Optional.empty();
+            }
+
+            Set<PlanNodeId> allSourceNodeIds = new HashSet<>(stageExecution.getFragment().getPartitionedSources());
+            stageExecution.getFragment().getRemoteSourceNodes().forEach(planNode -> allSourceNodeIds.add(planNode.getId()));
+
+            RemoteTask task = stageExecution.createTask(
+                    node,
+                    partition,
+                    attemptId,
+                    sinkBucketToPartitionMap,
+                    outputBuffers,
+                    taskSplits,
+                    allSourceNodeIds.stream()
+                            .collect(toImmutableListMultimap(Function.identity(), planNodeId -> Lifespan.taskWide())),
+                    allSourceNodeIds).orElseThrow(() -> new VerifyException("stage execution is expected to be active"));
+
+            partitionToRemoteTaskMap.put(partition, task);
+            runningTasks.put(task.getTaskId(), task);
+            runningNodes.put(task.getTaskId(), node);
+
+            if (taskFinishedFuture == null) {
+                taskFinishedFuture = SettableFuture.create();
+            }
+
+            taskLifecycleListener.taskCreated(stageExecution.getFragment().getId(), task);
+
+            task.addStateChangeListener(taskStatus -> updateTaskStatus(taskStatus, exchangeSinkInstanceHandle));
+            task.start();
+        }
+
+        if (taskFinishedFuture != null && !taskFinishedFuture.isDone()) {
+            blocked = taskFinishedFuture;
+        }
+    }
+
+    public synchronized boolean isFinished()
+    {
+        return failure == null &&
+                taskSource != null &&
+                taskSource.isFinished() &&
+                queuedPartitions.isEmpty() &&
+                finishedPartitions.containsAll(partitionToTaskDescriptorMap.keySet());
+    }
+
+    public void cancel()
+    {
+        close(false);
+    }
+
+    public void abort()
+    {
+        close(true);
+    }
+
+    private void fail(Throwable t)
+    {
+        synchronized (this) {
+            if (failure == null) {
+                failure = t;
+            }
+        }
+        close(true);
+    }
+
+    private void close(boolean abort)
+    {
+        boolean closed;
+        synchronized (this) {
+            closed = this.closed;
+            this.closed = true;
+        }
+        if (!closed) {
+            cancelRunningTasks(abort);
+            cancelBlockedFuture();
+            releaseAcquiredNode();
+            closeTaskSource();
+            closeSinkExchange();
```

Comment:
maybe:
```java
            AutoCloseableCloser closer = AutoCloseableCloser.create();
            closer.register(this::closeSinkExchange);
            closer.register(this::closeTaskSource);
            closer.register(this::releaseAcquiredNode);
            closer.register(this::cancelBlockedFuture);
            closer.register(() -> cancelRunningTasks(abort));
            try {
                closer.close();
            }
            catch (Exception e) {
                throw new RuntimeException(e);
            }
```

so we keep closing on exception?

Reply:
I was trying to implement the methods to avoid throwing an exception. Not sure if it makes sense to throw an exception when closing as it is not clear how to handle that exception? Should the query be failed when the close fails?

Reply:
Only if a failure in closing would affect the query results (e.g., incomplete results), otherwise, we should just log an error and ignore.

Reply:
This method is only executed when a query is finished (e.g.: cancelled, finished or failed)